### ISSUE TITLE
Allow SR users to use `livePreview` inline chat mode for diffs

### ISSUE THREAD

**Issue by meganrogge** [microsoft/vscode#202250]
Currently, by default, we enter the accessible diff view for screen reader users. 

@rperez030 has pointed out some reasons this might not be ideal. 
- one cannot move quickly between changes if there are a lot and there’s no way to go straight to the top
- he wants more context, being in the actual file without having to dismiss the widget, which discards the whole session
- there’s no indentation

If a user changes `inlineChat.mode` to `livePreview`, then an in editor experience can be obtained. When in screen reader mode, there's no way to go between the editor and the inline chat widget. When not in screen reader mode, `cmd+upArrow` / `cmd+downArrow` toggle between these. (We could probably fix this)

Even if I enter the editor via mouse click, when using arrow keys to navigate the in-editor div, I don't hear audio cues. @hediet would it be possible to support this or any thoughts?

**Comment by jrieken** [comment]
> If a user changes inlineChat.mode to livePreview, then an in editor experience can be obtained. When in screen reader mode, there's no way to go between the editor and the inline chat widget.

Just FYI that we are working to deprecate/remove `livePreview` and that the near future is `live` and `preview` only

**Comment by isidorn** [comment]
Based on @jrieken comment I suggest to re-asses this issue with the new behaviour (`live` and `preview`, especially the default one)

**Comment by meganrogge** [comment]
@jrieken which will be the default?

**Comment by jrieken** [comment]
I have removed the logic that makes all screenreader users use preview-mode. In return, the inline chat widget shows the accessible diff viewer when screenreader mode is enabled (and live-mode is used)

<img width="869" alt="Screenshot 2024-02-19 at 13 26 48" src="https://github.com/microsoft/vscode/assets/1794099/973a55c3-fc11-4675-96d2-a902d78cd990">

**Comment by jrieken** [comment]
To verify

* configure inline chat mode `live`
* let inline chat make a code change
* make sure you see an embedded accessible diff view
* ensure the diff viewer can be reached by `Tab` and lines can be visited with arrow keys

**Comment by meganrogge** [comment]
Thanks for testing @rperez030. 

The behavior now is that the added code appears both in the editor and in the inline chat widget's embedded accessible diff editor. 
 
There are a few issues @jrieken :
- it's not clear how to move focus between the editor and the inline chat widget using the keyboard. Is there a way to do this?
- if I click with the mouse into the editor and then navigate with arrow keys, there are no audio cue to indicate which lines have been inserted as I navigate

**Comment by jrieken** [comment]
> it's not clear how to move focus between the editor and the inline chat widget using the keyboard. Is there a way to do this?

From the input field you can use cmd+up/down to leave and re-enter the widget. Tho, when that's wanted it's probably best to disable the accessible diff viewer altogether (`inlineChat.accessibleDiffView`-setting). However, we haven't developed for that scenario as the accessible diff view is our "vehicle" for making code changes more accessible. 

> if I click with the mouse into the editor and then navigate with arrow keys, there are no audio cue to indicate which lines have been inserted as I navigate

Yes, that wasn't the goal. The embedded accessible diff view does this tho (see above)

> the deleted lines are not rendered in the editor

Yes, you need to toggle them with the "Show Changes" btn/command

**Comment by meganrogge** [comment]
> However, we haven't developed for that scenario as the accessible diff view is our "vehicle" for making code changes more accessible.

I think this issue should be reopened as the ask is to have a way to navigate the changes in the editor - not the accessible diff viewer - in an accessible way.

**Comment by rperez030** [comment]
> I think this issue should be reopened as the ask is to have a way to navigate the changes in the editor - not the accessible diff viewer - in an accessible way.

That is correct. The accessible diff view does not offer equivalent functionality when it comes to reviewing code suggestions from Copilot. It is effective when reviewing changes in a pull request, for example, where one can explore the underlying code and use the accessible diff view, if needed, to examine a single change. However, when Copilot suggestions are provided only through the accessible view, there is no way for me to obtain context from the code it affects or navigate efficiently when the change affects a number of lines. We received similar feedback from one of the participants at the latest Microsoft WBU Summit, which convinced me that it wasn't just me.

Reviewing code in the editor should be accessible by default. The accessible diff view should be available on demand to review an individual change if needed.

**Comment by jrieken** [comment]
> The accessible diff view does not offer equivalent functionality when it comes to reviewing code suggestions from Copilot.

Sorry, but I need to correct this. You are not reviewing code suggestions but changes that have already been made. This is why we use the familiar diff view. It doesn't mean the diff viewer is done but we need to agree on the overall approach of using it (just like we do it for editable or readonly diff editors) Last, the changes are not just in the diff viewer but also in the editor - the diff viewer has some context, the editor the whole context.

**Comment by rperez030** [comment]
> Sorry, but I need to correct this. You are not reviewing code suggestions but changes that have already been made. This is why we use the familiar diff view. It doesn't mean the diff viewer is done but we need to agree on the overall approach of using it (just like we do it for editable or readonly diff editors) Last, the changes are not just in the diff viewer but also in the editor - the diff viewer has some context, the editor the whole context.

@jrieken thank you for the clarification. Yes, I understand the changes have already been made in the code, but the user still can reject them, right? If that is the case, it is best to have the whole context accessible when assessing these changes. These are the reasons why I believe access to the editor is necessary.

* Context: The few lines of code contained in the diff viewer may not be enough to fully understand the changes, particularly if one is not familiar with the code.
* Ease off reading: reading code in the editor is just easier. One cannot only examine the code  line by line, but also character by character.
* Indentation: the diff viewer does not expose indentation to screen readers, and even if it did, screen readers do not expect to have to account for indentation when reviewing items in a list. Indentation is essential to understand code written in Python, for example.
* Efficient navigation between changes. a code suggestion from copilot may contain changes that are a number of lines apart. In the diff viewer, one must navigate the entire list line by line to read all the changes, while in the editor it should be possible to press Alt +F5 or Shift +f5 to move the cursor between next and previous change respectively. I don't know if that command would work in the context of copilot suggestions, but it does in the context of comparing files.

Does it make sense?

**Comment by jrieken** [comment]
> one must navigate the entire list line by line to read all the changes, while in the editor it should be possible to press Alt +F5 or Shift +f5 to move the cursor between next and previous change respectively.

For inline chat this is F7 and Shift+F7 - it will visit all major regions of changes (hunks) which then contain smaller, colocated changes. 

I wonder if indentation, ease of reading, and context are general issues with the accessible diff view or not? So, when doing reviews of your own or incoming changes do you think the same shortcomings apply to the accessible diff view? I am asking because inline chat uses very similar visuals as the diff editor does, so it appears natural to also have a similar accessibility component. That doesn't mean the accessible diff view is perfect but I am also not convinced that it is the totally wrong tool for the job. 


> If that is the case, it is best to have the whole context accessible when assessing these changes. These are the reasons why I believe access to the editor is necessary

Assuming you have the accessible diff view disabled (`inlineChat.accessibleDiffView`-setting) how should this work? Do you just want a signal, like an audio cue, that says a line has been AI authored/changed? Is that enough or do you want it to read the exact changes. Think of long lines and small changes and how you want to perceive it.

**Comment by jrieken** [comment]
ping @rperez030

**Comment by rperez030** [comment]
hi @jrieken . I'm very sorry I have not followed up with you on this. I'm in meetings for the rest of the day but I will give you an answer tomorrow. One thing I can say is that they key-bindings for moving in and out of the chat are configure so that they don't work with screen reader mode enabled. i will change the rules, test the experience and let you know how far i can go and what the experience is on my en.

again, my apologies for the silence.
